core/state/statedb.go

```diff
@@ -724,33 +724,55 @@ func (s *StateDB) clearJournalAndRefund() {
 	s.validRevisions = s.validRevisions[:0] // Snapshots can be created without journal entires
 }
 
+type ModifiedAccount struct {
+	Account
+	Storage
+}
+type StateChanges struct {
+	ModifiedAccounts map[common.Address]ModifiedAccount
+}
+
 // Commit writes the state to the underlying in-memory trie database.
-func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
+func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, StateChanges, error) {
 	// Finalize any pending changes and merge everything into the tries
 	s.IntermediateRoot(deleteEmptyObjects)
 
+	modifiedAccounts := StateChanges{
+		ModifiedAccounts: make(map[common.Address]ModifiedAccount),
+	}
+
 	// Commit objects to the trie, measuring the elapsed time
 	for addr := range s.stateObjectsDirty {
+		dirtyStateObject := s.stateObjects[addr]
+		modifiedAccount := ModifiedAccount{Account: dirtyStateObject.data} //state_object.data is the account
+
 		if obj := s.stateObjects[addr]; !obj.deleted {
 			// Write any contract code associated with the state object
 			if obj.code != nil && obj.dirtyCode {
 				s.db.TrieDB().InsertBlob(common.BytesToHash(obj.CodeHash()), obj.code)
 				obj.dirtyCode = false
 			}
+
+			// Add the dirty storage to the modifiedAccounts map before it's committed (and flushed from in-memory)
+			modifiedAccount.Storage = dirtyStateObject.dirtyStorage
+
 			// Write any storage changes in the state object to its storage trie
 			if err := obj.CommitTrie(s.db); err != nil {
-				return common.Hash{}, err
+				return common.Hash{}, StateChanges{}, err
 			}
 		}
+
+		modifiedAccounts.ModifiedAccounts[addr] = modifiedAccount
```

Comment:
Just to make sure I understand - is it correct that a deleted account will result in the payload mapping the address to an empty `ModifiedAccount`?

Reply:
I think that's correct that `Commit` would return a mapping from an address to an empty `ModifiedAccount`. I recently made a change to not return Payload with empty `ModifiedAccounts`, so that we're not spamming the subscribers with empty payloads. But I'd be happy to think that through a bit more.